### PR TITLE
planner: support column range partition pruning for builtin function IN

### LINKED ISSUE #26739: operator 'in' partition pruning doesn't take effect on date partition key
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
CREATE TABLE t1 (
`id` bigint(20)  NOT NULL AUTO_INCREMENT,
`dt` date,
PRIMARY KEY (`id`,`dt`)
)
PARTITION BY RANGE COLUMNS(dt) (
PARTITION `p20201125` VALUES LESS THAN ("20201126"),
PARTITION `p20201126` VALUES LESS THAN ("20201127"),
PARTITION `p20201127` VALUES LESS THAN ("20201128"),
PARTITION `p20201128` VALUES LESS THAN ("20201129"),
PARTITION `p20201129` VALUES LESS THAN ("20201130")
);

explain select count(1) from t1 where dt in ('2020-11-27','2020-11-28');
```

### 2. What did you expect to see? (Required)

```
+-------------------------------------+----------+-----------+------------------------------------------------------+------------------------------------------------------------------------+
| id                                  | estRows  | task      | access object                                        | operator info                                                          |
+-------------------------------------+----------+-----------+------------------------------------------------------+------------------------------------------------------------------------+
| HashAgg_18                          | 1.00     | root      |                                                      | funcs:count(Column#5)->Column#4                                        |
| └─PartitionUnion_19                 | 5.00     | root      |                                                      |                                                                        |
|   ├─StreamAgg_99                    | 1.00     | root      |                                                      | funcs:count(Column#22)->Column#5                                       |
|   │ └─IndexReader_100               | 1.00     | root      |                                                      | index:StreamAgg_81                                                     |
|   │   └─StreamAgg_81                | 1.00     | cop[tikv] |                                                      | funcs:count(1)->Column#22                                              |
|   │     └─Selection_98              | 20.00    | cop[tikv] |                                                      | in(test.t1.dt, 2020-11-27 00:00:00.000000, 2020-11-28 00:00:00.000000) |
|   │       └─IndexFullScan_97        | 10000.00 | cop[tikv] | table:t1, partition:p20201127, index:PRIMARY(id, dt) | keep order:false, stats:pseudo                                         |
|   └─StreamAgg_128                   | 1.00     | root      |                                                      | funcs:count(Column#28)->Column#5                                       |
|     └─IndexReader_129               | 1.00     | root      |                                                      | index:StreamAgg_110                                                    |
|       └─StreamAgg_110               | 1.00     | cop[tikv] |                                                      | funcs:count(1)->Column#28                                              |
|         └─Selection_127             | 20.00    | cop[tikv] |                                                      | in(test.t1.dt, 2020-11-27 00:00:00.000000, 2020-11-28 00:00:00.000000) |
|           └─IndexFullScan_126       | 10000.00 | cop[tikv] | table:t1, partition:p20201128, index:PRIMARY(id, dt) | keep order:false, stats:pseudo                                         |
+-------------------------------------+----------+-----------+------------------------------------------------------+------------------------------------------------------------------------+
```

### 3. What did you see instead (Required)

```
+-------------------------------------+----------+-----------+------------------------------------------------------+------------------------------------------------------------------------+
| id                                  | estRows  | task      | access object                                        | operator info                                                          |
+-------------------------------------+----------+-----------+------------------------------------------------------+------------------------------------------------------------------------+
| HashAgg_18                          | 1.00     | root      |                                                      | funcs:count(Column#5)->Column#4                                        |
| └─PartitionUnion_19                 | 5.00     | root      |                                                      |                                                                        |
|   ├─StreamAgg_41                    | 1.00     | root      |                                                      | funcs:count(Column#10)->Column#5                                       |
|   │ └─IndexReader_42                | 1.00     | root      |                                                      | index:StreamAgg_23                                                     |
|   │   └─StreamAgg_23                | 1.00     | cop[tikv] |                                                      | funcs:count(1)->Column#10                                              |
|   │     └─Selection_40              | 20.00    | cop[tikv] |                                                      | in(test.t1.dt, 2020-11-27 00:00:00.000000, 2020-11-28 00:00:00.000000) |
|   │       └─IndexFullScan_39        | 10000.00 | cop[tikv] | table:t1, partition:p20201125, index:PRIMARY(id, dt) | keep order:false, stats:pseudo                                         |
|   ├─StreamAgg_70                    | 1.00     | root      |                                                      | funcs:count(Column#16)->Column#5                                       |
|   │ └─IndexReader_71                | 1.00     | root      |                                                      | index:StreamAgg_52                                                     |
|   │   └─StreamAgg_52                | 1.00     | cop[tikv] |                                                      | funcs:count(1)->Column#16                                              |
|   │     └─Selection_69              | 20.00    | cop[tikv] |                                                      | in(test.t1.dt, 2020-11-27 00:00:00.000000, 2020-11-28 00:00:00.000000) |
|   │       └─IndexFullScan_68        | 10000.00 | cop[tikv] | table:t1, partition:p20201126, index:PRIMARY(id, dt) | keep order:false, stats:pseudo                                         |
|   ├─StreamAgg_99                    | 1.00     | root      |                                                      | funcs:count(Column#22)->Column#5                                       |
|   │ └─IndexReader_100               | 1.00     | root      |                                                      | index:StreamAgg_81                                                     |
|   │   └─StreamAgg_81                | 1.00     | cop[tikv] |                                                      | funcs:count(1)->Column#22                                              |
|   │     └─Selection_98              | 20.00    | cop[tikv] |                                                      | in(test.t1.dt, 2020-11-27 00:00:00.000000, 2020-11-28 00:00:00.000000) |
|   │       └─IndexFullScan_97        | 10000.00 | cop[tikv] | table:t1, partition:p20201127, index:PRIMARY(id, dt) | keep order:false, stats:pseudo                                         |
|   ├─StreamAgg_128                   | 1.00     | root      |                                                      | funcs:count(Column#28)->Column#5                                       |
|   │ └─IndexReader_129               | 1.00     | root      |                                                      | index:StreamAgg_110                                                    |
|   │   └─StreamAgg_110               | 1.00     | cop[tikv] |                                                      | funcs:count(1)->Column#28                                              |
|   │     └─Selection_127             | 20.00    | cop[tikv] |                                                      | in(test.t1.dt, 2020-11-27 00:00:00.000000, 2020-11-28 00:00:00.000000) |
|   │       └─IndexFullScan_126       | 10000.00 | cop[tikv] | table:t1, partition:p20201128, index:PRIMARY(id, dt) | keep order:false, stats:pseudo                                         |
|   └─StreamAgg_157                   | 1.00     | root      |                                                      | funcs:count(Column#34)->Column#5                                       |
|     └─IndexReader_158               | 1.00     | root      |                                                      | index:StreamAgg_139                                                    |
|       └─StreamAgg_139               | 1.00     | cop[tikv] |                                                      | funcs:count(1)->Column#34                                              |
|         └─Selection_156             | 20.00    | cop[tikv] |                                                      | in(test.t1.dt, 2020-11-27 00:00:00.000000, 2020-11-28 00:00:00.000000) |
|           └─IndexFullScan_155       | 10000.00 | cop[tikv] | table:t1, partition:p20201129, index:PRIMARY(id, dt) | keep order:false, stats:pseudo                                         |
+-------------------------------------+----------+-----------+------------------------------------------------------+------------------------------------------------------------------------+
```

### 4. What is your TiDB version? (Required)
v5.1.0

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: close #26739

Problem Summary: planner: support column range partition pruning for builtin function IN

### What is changed and how it works?

Convert IN to EQ and then prune them.

### Check List

Tests 

- [x] Unit test

### Release note

```release-note
planner: support column range partition pruning for builtin function IN
```

### CODE REVIEW

**Review by rebelice — APPROVED** [review]

**Comment by sultan8252** [comment]
👍

**Comment by qw4990** [comment]
PTAL @Reminiscent

File: planner/core/partition_pruner_test.go [hunk 1]
```diff
@@ -86,6 +86,83 @@ func (s *testPartitionPruneSuit) TestHashPartitionPruner(c *C) {
...
+	tk.MustExec("use test_range_col_in")
+	tk.MustExec(`set @@session.tidb_enable_list_partition = 1`)
+	tk.MustExec("set @@session.tidb_partition_prune_mode='static'")
+
+	// case in issue-26739
+	tk.MustExec(`CREATE TABLE t1 (
+		id bigint(20)  NOT NULL AUTO_INCREMENT,
+		dt date,
+		PRIMARY KEY (id,dt))
+		PARTITION BY RANGE COLUMNS(dt) (
```
Comment (Reminiscent) [inline planner/core/partition_pruner_test.go:102]: Please add more tests about `PARTITION BY RANGE COLUMNS(a,d,c)`(partition by more columns). Rest LGTM

**Review by Reminiscent — APPROVED** [review]

File: planner/core/partition_pruner_test.go (hunk 1 above)
Comment (qw4990) [reply planner/core/partition_pruner_test.go:102]: Multiple-column range partitioning is not supported now~:
```
mysql> create table tx (
    -> a int, 
    -> b int,
    -> c int
    -> ) partition by range columns (a, b, c) (
    -> partition p0 values less than (10, 10, 10),
    -> partition p1 values less than (20, 20, 20)
    -> );
Query OK, 0 rows affected, 1 warning (0.01 sec)

mysql> show warnings;
+---------+------+---------------------------------------------------------+
| Level   | Code | Message                                                 |
+---------+------+---------------------------------------------------------+
| Warning | 8200 | Unsupported partition type RANGE, treat as normal table |
+---------+------+---------------------------------------------------------+
1 row in set (0.00 sec)

mysql> show create table tx;
+-------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                       |
+-------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tx    | CREATE TABLE `tx` (
  `a` int(11) DEFAULT NULL,
  `b` int(11) DEFAULT NULL,
  `c` int(11) DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
+-------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

```

**Pull request merged**